config/impl.go

```diff
@@ -21,57 +21,94 @@
 package config
 
 import (
+	"context"
 	"fmt"
+	"reflect"
+	"strings"
+
+	"go.uber.org/fx"
+
 	"go.ketch.com/lib/orlop/v2/env"
 	"go.ketch.com/lib/orlop/v2/errors"
 	"go.ketch.com/lib/orlop/v2/service"
-	"go.uber.org/fx"
-	"reflect"
-	"strings"
 )
 
 type Config interface {
 	Options() fx.Option
 }
 
 type providerImpl struct {
+	configs map[string]any
 	environ env.Environ
+	prefix  service.Name
 }
 
-func New(environ env.Environ) Provider {
+func New(p Params) Provider {
+	configs := make(map[string]any)
+	for _, c := range p.Defs {
+		configs[c.Name] = c.Config
+	}
+
 	return &providerImpl{
-		environ: environ,
+		configs: configs,
+		environ: p.Environ,
+		prefix:  p.Prefix,
 	}
 }
 
-func (s *providerImpl) Load(cfg Config) error {
-	fields, err := reflectStruct([]string{}, cfg)
-	if err != nil {
-		return err
+func (s *providerImpl) Get(_ context.Context, service string) (any, error) {
+	if cfg, ok := s.configs[service]; ok {
+		return cfg, nil
 	}
 
-	for name, field := range fields {
-		if v := s.environ.Getenv(name); len(v) > 0 {
-			if err = field.set(field.v, v); err != nil {
-				return errors.Wrapf(err, "failed to set field '%s' with value '%s'", name, v)
-			}
-		} else if field.tag.DefaultValue != nil {
-			if err = field.set(field.v, *field.tag.DefaultValue); err != nil {
-				return errors.Wrapf(err, "failed to set field '%s' with value '%s'", name, v)
+	return nil, errors.Errorf("%s config not found", service)
+}
+
+func (s *providerImpl) List(_ context.Context) ([]string, error) {
+	var vars []string
+	for k, v := range s.configs {
+		vs, err := getVariablesFromConfig(s.prefix, k, v)
+		if err != nil {
+			return nil, err
+		}
+
+		vars = append(vars, vs...)
+	}
+
+	return vars, nil
+}
+
+func (s *providerImpl) Load(_ context.Context) error {
+	for k, v := range s.configs {
+		fields, err := reflectStruct([]string{}, v)
+		if err != nil {
+			return err
+		}
+
+		for name, field := range fields {
+			keyName := strings.Join([]string{k, name}, "_")
+			if v := s.environ.Getenv(keyName); len(v) > 0 {
+				if err = field.set(field.v, v); err != nil {
+					return errors.Wrapf(err, "failed to set field '%s' with value '%s'", name, v)
+				}
+			} else if field.tag.DefaultValue != nil {
+				if err = field.set(field.v, *field.tag.DefaultValue); err != nil {
+					return errors.Wrapf(err, "failed to set field '%s' with value '%s'", name, v)
+				}
+			} else if field.tag.Required {
+				return errors.Errorf("%s required", name)
 			}
-		} else if field.tag.Required {
-			return errors.Errorf("%s required", name)
 		}
 	}
 
 	return nil
 }
 
-// GetVariablesFromConfig returns the environment variables from the given config object
-func GetVariablesFromConfig(prefix service.Name, cfg Config) ([]string, error) {
+// getVariablesFromConfig returns the environment variables from the given config object
+func getVariablesFromConfig(prefix service.Name, service string, cfg any) ([]string, error) {
```

Comment:
If this is not used anywhere else, you can make a member function of the struct and would then have prefix available there.